From a0726db0aec7556b1e9014e568d970aa59293728 Mon Sep 17 00:00:00 2001
From: Szehon Ho <szehon@cloudera.com>
Date: Tue, 24 Jun 2014 18:06:53 -0700
Subject: [PATCH 0011/1164] CLOUDERA-BUILD: Build Hive Spark Branch on CDH5

CDH-20058: Build Hive 0.13 branch on CDH5 components

Conflicts:
	ant/pom.xml
	beeline/pom.xml
	cli/pom.xml
	common/pom.xml
	contrib/pom.xml
	hbase-handler/pom.xml
	hcatalog/core/pom.xml
	hcatalog/core/src/main/java/org/apache/hcatalog/mapreduce/MultiOutputFormat.java
	hcatalog/core/src/test/java/org/apache/hcatalog/mapreduce/TestMultiOutputFormat.java
	hcatalog/hcatalog-pig-adapter/pom.xml
	hcatalog/hcatalog-pig-adapter/src/main/java/org/apache/hcatalog/pig/HCatLoader.java
	hcatalog/hcatalog-pig-adapter/src/main/java/org/apache/hcatalog/pig/HCatStorer.java
	hcatalog/pom.xml
	hcatalog/server-extensions/pom.xml
	hcatalog/storage-handlers/hbase/pom.xml
	hcatalog/streaming/pom.xml
	hcatalog/webhcat/java-client/pom.xml
	hcatalog/webhcat/svr/pom.xml
	hwi/pom.xml
	itests/custom-serde/pom.xml
	itests/hcatalog-unit/pom.xml
	itests/hive-unit/pom.xml
	itests/pom.xml
	itests/qtest/pom.xml
	itests/test-serde/pom.xml
	itests/util/pom.xml
	itests/util/src/main/java/org/apache/hadoop/hive/ql/hooks/VerifyOverriddenConfigsHook.java
	jdbc/pom.xml
	metastore/pom.xml
	odbc/pom.xml
	packaging/pom.xml
	pom.xml
	ql/pom.xml
	serde/pom.xml
	service/pom.xml
	shims/0.20/pom.xml
	shims/0.20S/pom.xml
	shims/0.23/pom.xml
	shims/0.23/src/main/java/org/apache/hadoop/hive/shims/Hadoop23Shims.java
	shims/aggregator/pom.xml
	shims/common-secure/pom.xml
	shims/common/pom.xml
	shims/pom.xml
	testutils/pom.xml

CLOUDERA-BUILD: CDH-16128: HiveServer2 does not provide the MR2 jobs IDs by default

Conflicts:
	common/src/main/resources/hive-log4j.properties

(cherry picked from commit c8b7dc21d5020205bf16c17157e376d9c93bf156)

CLOUDERA-BUILD: CDH-19733: Set SENTRY_HOME from metastore script

(cherry picked from commit 1bbc4f9978ed40b26c8ea6606e54fb1976ef45d9)
(cherry picked from commit d22ff3308ea5b43ff1c494d000b3ca4e36463a65)

CLOUDERA-BUILD: CDH-16414 - Hadoop23 Shim uses Java 7 only constructor

(cherry picked from commit 062bcceb8ea9f5c3ede3652ce03f0556490ae26b)
(cherry picked from commit 9be7140621aa7797abe2649002c2760fc2a038d3)

CLOUDERA-BUILD: CDH-17152: Restore default RC file format serde to ColumnarSerde

(cherry picked from commit 1304088fbd3f5c8f92b64c5db4d37b85ebf26ce6)

Conflicts:
	conf/hive-default.xml.template

(cherry picked from commit 8e005c4976b7145c10c6f97fa2c4b4ca7a0a5eab)

Conflicts:
	common/src/java/org/apache/hadoop/hive/conf/HiveConf.java
	conf/hive-default.xml.template

CLOUDERA-BUILD: CDH-17875: Deprecate HiveServer2 SSL configuration parameter

(cherry picked from commit c98684a56c0944dd71c5d4d3bbc6d660bf4da8b2)

Conflicts:
	itests/hive-unit/src/test/java/org/apache/hive/jdbc/TestSSL.java

(cherry picked from commit 31096bc355d1050e3ebff19c905c8c722ec6d471)

Conflicts:
	common/src/java/org/apache/hadoop/hive/conf/HiveConf.java

CLOUDERA-BUILD: CDH-16507: Fix Hive MR2 test failures

(cherry picked from commit 5298668593f3818233b39c59d1351a138952ec24)

CLOUDERA-BUILD: CDH-16001 - QTest module under hadoop-1 is missing hadoop-common

(cherry picked from commit 725155e2ea64f392a215b6991aca8d500dc056af)

CLOUDERA-BUILD: CDH-16510 - Fix Hive MR1 test failures

(cherry picked from commit 114165a67d19815b1915454257345ed886c2f054)
(cherry picked from commit 003905152884401b8d18452b1616790c1f27833b)

CLOUDERA-BUILD: CDH-21747: Set the default authorization factory to disable built-in authorization

(cherry picked from commit 1cc395263f16c7282ae37b266d7ab21f9396cdc0)

Conflicts:
	common/src/java/org/apache/hadoop/hive/conf/HiveConf.java
	data/conf/hive-site.xml

CLOUDERA-BUILD: CDH-21747: Set the default authorization factory to disable built-in authorization

(cherry picked from commit 523f8358bbe097e51f7044b2672ade9061b36a87)
(cherry picked from commit 8cb92b79f97ab18a0ad7a54a55d4f071e0acbff7)

Conflicts:
	accumulo-handler/pom.xml
	ant/pom.xml
	beeline/pom.xml
	cli/pom.xml
	common/pom.xml
	contrib/pom.xml
	hbase-handler/pom.xml
	hcatalog/core/pom.xml
	hcatalog/hcatalog-pig-adapter/pom.xml
	hcatalog/pom.xml
	hcatalog/server-extensions/pom.xml
	hcatalog/streaming/pom.xml
	hcatalog/webhcat/java-client/pom.xml
	hcatalog/webhcat/svr/pom.xml
	hwi/pom.xml
	itests/custom-serde/pom.xml
	itests/hcatalog-unit/pom.xml
	itests/hive-minikdc/pom.xml
	itests/hive-unit-hadoop2/pom.xml
	itests/hive-unit/pom.xml
	itests/pom.xml
	itests/qtest-spark/pom.xml
	itests/qtest/pom.xml
	itests/test-serde/pom.xml
	itests/util/pom.xml
	jdbc/pom.xml
	metastore/pom.xml
	odbc/pom.xml
	packaging/pom.xml
	pom.xml
	ql/pom.xml
	ql/src/java/org/apache/hadoop/hive/ql/io/IOContext.java
	serde/pom.xml
	service/pom.xml
	shims/0.20S/pom.xml
	shims/0.23/pom.xml
	shims/0.23/src/main/java/org/apache/hadoop/hive/shims/Hadoop23Shims.java
	shims/aggregator/pom.xml
	shims/common/pom.xml
	shims/pom.xml
	shims/scheduler/pom.xml
	spark-client/pom.xml
	testutils/pom.xml

CLOUDERA-BUILD: hard code spark.version in tarball until CDH-24361 is fixed

(cherry picked from commit 266330a94949f944c549706d908e8832b648b054)

CLOUDERA-BUILD: Use new MRVersion class and fix SPARK_SUBMIT_CLASSPATH

(cherry picked from commit 0370469e715a9acd578c5eafa8d5446720ceb809)

CLOUDERA-BUILD: add apache snapshots repo

(cherry picked from commit 812209a599a6082122fbb6aa3f893dd7d7b3a598)

CLOUDERA-BUILD: Fixes to allow build to work with hadoop-1 for tests

(cherry picked from commit 0bfd450327b33daf5aba82b7e17fa1debb883e86)

CDH-17697 Analyze hive unit test failures for MR1

Conflicts:
	ql/pom.xml
	serde/pom.xml
	hcatalog/core/pom.xml
	hcatalog/storage-handlers/hbase/pom.xml
	itests/hcatalog-unit/pom.xml

(cherry picked from commit e5cd0fa9b55aca2eb57883bbe27586716f20e093)

Conflicts:
	hcatalog/core/pom.xml
	hcatalog/storage-handlers/hbase/pom.xml
	hwi/pom.xml
	itests/hcatalog-unit/pom.xml
	ql/pom.xml

(cherry picked from commit dce1606551208b015080a5bf9ed610743f0202d5)

CLOUDERA-BUILD: Fix additional MR1 Use cases

(cherry picked from commit 10cd9cbe00082cbc8ba3ab507dfedbe3662e78e9)
(cherry picked from commit cba3a4aa0bf911e547f8b462d91aad8360f567f6)
---
 accumulo-handler/pom.xml                           |    2 +-
 ant/pom.xml                                        |    2 +-
 beeline/pom.xml                                    |    5 +-
 bin/ext/metastore.sh                               |   15 ++
 cli/pom.xml                                        |    5 +-
 common/pom.xml                                     |   18 ++-
 .../java/org/apache/hadoop/hive/conf/HiveConf.java |    6 +-
 common/src/main/resources/hive-log4j.properties    |   11 +-
 common/src/test/resources/core-site.xml            |    4 +
 contrib/pom.xml                                    |   18 ++-
 data/conf/hive-site.xml                            |    8 +-
 hbase-handler/pom.xml                              |   11 +-
 hcatalog/core/pom.xml                              |  102 ++++++++------
 .../hive/hcatalog/mapreduce/MultiOutputFormat.java |    5 +-
 .../hcatalog/mapreduce/TestMultiOutputFormat.java  |   13 +-
 hcatalog/hcatalog-pig-adapter/pom.xml              |   12 +-
 .../org/apache/hive/hcatalog/pig/HCatLoader.java   |    2 +-
 .../org/apache/hive/hcatalog/pig/HCatStorer.java   |    2 +-
 hcatalog/pom.xml                                   |   12 +-
 hcatalog/server-extensions/pom.xml                 |   11 +-
 hcatalog/streaming/pom.xml                         |    2 +-
 hcatalog/webhcat/java-client/pom.xml               |    5 +-
 hcatalog/webhcat/svr/pom.xml                       |   36 ++---
 hwi/pom.xml                                        |   11 +-
 itests/custom-serde/pom.xml                        |    5 +-
 itests/hcatalog-unit/pom.xml                       |   89 ++++++------
 itests/hive-jmh/pom.xml                            |    2 +-
 itests/hive-minikdc/pom.xml                        |    6 +-
 itests/hive-unit-hadoop2/pom.xml                   |    2 +-
 itests/hive-unit/pom.xml                           |   17 ++-
 .../test/java/org/apache/hive/jdbc/TestSSL.java    |   24 +++-
 itests/pom.xml                                     |    4 +-
 itests/qtest-spark/pom.xml                         |    2 +-
 itests/qtest/pom.xml                               |   29 +++-
 itests/test-serde/pom.xml                          |    5 +-
 itests/util/pom.xml                                |    5 +-
 .../hive/ql/hooks/VerifyOverriddenConfigsHook.java |    2 +-
 jdbc/pom.xml                                       |    5 +-
 metastore/pom.xml                                  |   18 ++-
 odbc/pom.xml                                       |    2 +-
 packaging/pom.xml                                  |   12 +-
 pom.xml                                            |  141 ++++++++++++++++---
 ql/pom.xml                                         |   14 +-
 .../apache/hadoop/hive/ql/exec/mr/JobDebugger.java |    5 +-
 .../org/apache/hadoop/hive/ql/io/IOContext.java    |    1 -
 ...RestrictedHiveAuthorizationTaskFactoryImpl.java |  143 ++++++++++++++++++++
 serde/pom.xml                                      |   57 ++++----
 service/pom.xml                                    |   18 ++-
 shims/0.20S/pom.xml                                |   11 +-
 shims/0.23/pom.xml                                 |    2 +-
 .../apache/hadoop/hive/shims/Hadoop23Shims.java    |   83 +++++++++---
 shims/aggregator/pom.xml                           |    4 +-
 shims/common/pom.xml                               |    8 +-
 shims/pom.xml                                      |    4 +-
 shims/scheduler/pom.xml                            |    2 +-
 spark-client/pom.xml                               |    3 +-
 testutils/pom.xml                                  |    2 +-
 57 files changed, 778 insertions(+), 267 deletions(-)
 create mode 100644 ql/src/java/org/apache/hadoop/hive/ql/parse/authorization/RestrictedHiveAuthorizationTaskFactoryImpl.java

diff --git a/accumulo-handler/pom.xml b/accumulo-handler/pom.xml
index b9c794f..b2a05b2 100644
--- a/accumulo-handler/pom.xml
+++ b/accumulo-handler/pom.xml
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
diff --git a/ant/pom.xml b/ant/pom.xml
index ee9bd7a..fe3bade 100644
--- a/ant/pom.xml
+++ b/ant/pom.xml
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
diff --git a/beeline/pom.xml b/beeline/pom.xml
index 2e4bcac..c844c17 100644
--- a/beeline/pom.xml
+++ b/beeline/pom.xml
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
@@ -116,6 +116,9 @@
   <profiles>
     <profile>
       <id>hadoop-1</id>
+      <activation>
+        <activeByDefault>false</activeByDefault>
+      </activation>
       <dependencies>
         <dependency>
           <groupId>org.apache.hadoop</groupId>
diff --git a/bin/ext/metastore.sh b/bin/ext/metastore.sh
index 22b2d5d..b294942 100644
--- a/bin/ext/metastore.sh
+++ b/bin/ext/metastore.sh
@@ -24,6 +24,21 @@ metastore() {
   fi
   JAR=${HIVE_LIB}/hive-service-*.jar
 
+  # Set SENTRY_HOME if possible and add Sentry jars to classpath
+  if [[ -z "$SENTRY_HOME" ]]
+  then
+    if [[ -d ${HIVE_HOME}/../sentry ]]
+    then
+      export SENTRY_HOME=`readlink -m ${HIVE_HOME}/../sentry`
+    fi
+  fi
+  if [[ -n "$SENTRY_HOME" ]]
+  then
+    for f in ${SENTRY_HOME}/lib/*.jar; do
+      export HADOOP_CLASSPATH=${HADOOP_CLASSPATH}:${f}
+    done
+  fi
+
   # hadoop 20 or newer - skip the aux_jars option and hiveconf
 
   export HADOOP_OPTS="$HIVE_METASTORE_HADOOP_OPTS $HADOOP_OPTS"
diff --git a/cli/pom.xml b/cli/pom.xml
index 047472e..a4534f5 100644
--- a/cli/pom.xml
+++ b/cli/pom.xml
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
@@ -114,6 +114,9 @@
   <profiles>
     <profile>
       <id>hadoop-1</id>
+      <activation>
+        <activeByDefault>false</activeByDefault>
+      </activation>
       <dependencies>
         <dependency>
           <groupId>org.apache.hadoop</groupId>
diff --git a/common/pom.xml b/common/pom.xml
index f24e5c4..d766154 100644
--- a/common/pom.xml
+++ b/common/pom.xml
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
@@ -88,11 +88,21 @@
       <version>${junit.version}</version>
       <scope>test</scope>
     </dependency>
+    <!-- Following have only one version, which is cdh.hadoop.version -->
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-common</artifactId>
+      <version>${hadoop-23.version}</version>
+      <optional>true</optional>
+    </dependency>
   </dependencies>
 
   <profiles>
     <profile>
       <id>hadoop-1</id>
+      <activation>
+        <activeByDefault>false</activeByDefault>
+      </activation>
       <dependencies>
         <dependency>
           <groupId>org.apache.hadoop</groupId>
@@ -107,12 +117,6 @@
       <dependencies>
         <dependency>
           <groupId>org.apache.hadoop</groupId>
-          <artifactId>hadoop-common</artifactId>
-          <version>${hadoop-23.version}</version>
-          <optional>true</optional>
-        </dependency>
-        <dependency>
-          <groupId>org.apache.hadoop</groupId>
           <artifactId>hadoop-mapreduce-client-core</artifactId>
           <version>${hadoop-23.version}</version>
           <optional>true</optional>
diff --git a/common/src/java/org/apache/hadoop/hive/conf/HiveConf.java b/common/src/java/org/apache/hadoop/hive/conf/HiveConf.java
index 7c7ff43..bf0831f 100644
--- a/common/src/java/org/apache/hadoop/hive/conf/HiveConf.java
+++ b/common/src/java/org/apache/hadoop/hive/conf/HiveConf.java
@@ -105,6 +105,8 @@ public void setSparkConfigUpdated(boolean isSparkConfigUpdated) {
     for (ConfVars confVar : ConfVars.values()) {
       vars.put(confVar.varname, confVar);
     }
+    Configuration.addDeprecation("hive.server2.enable.impersonation", "hive.server2.enable.doAs");
+    Configuration.addDeprecation("hive.server2.enable.SSL", "hive.server2.use.SSL");
   }
 
   /**
@@ -775,7 +777,7 @@ public void setSparkConfigUpdated(boolean isSparkConfigUpdated) {
 
     // default serde for rcfile
     HIVEDEFAULTRCFILESERDE("hive.default.rcfile.serde",
-        "org.apache.hadoop.hive.serde2.columnar.LazyBinaryColumnarSerDe",
+        "org.apache.hadoop.hive.serde2.columnar.ColumnarSerDe",
         "The default SerDe Hive will use for the RCFile format"),
 
     HIVEDEFAULTSERDE("hive.default.serde",
@@ -1528,7 +1530,7 @@ public void setSparkConfigUpdated(boolean isSparkConfigUpdated) {
         "of the table. Note that the default gives the creator of a table no access to the\n" +
         "table (but see HIVE-8067)."),
     HIVE_AUTHORIZATION_TASK_FACTORY("hive.security.authorization.task.factory",
-        "org.apache.hadoop.hive.ql.parse.authorization.HiveAuthorizationTaskFactoryImpl",
+        "org.apache.hadoop.hive.ql.parse.authorization.RestrictedHiveAuthorizationTaskFactoryImpl",
         "Authorization DDL task factory implementation"),
 
     // if this is not set default value is set during config initialization
diff --git a/common/src/main/resources/hive-log4j.properties b/common/src/main/resources/hive-log4j.properties
index 14fa725..3f84af8 100644
--- a/common/src/main/resources/hive-log4j.properties
+++ b/common/src/main/resources/hive-log4j.properties
@@ -16,7 +16,7 @@
 
 # Define some default values that can be overridden by system properties
 hive.log.threshold=ALL
-hive.root.logger=INFO,DRFA
+hive.root.logger=WARN,DRFA
 hive.log.dir=${java.io.tmpdir}/${user.name}
 hive.log.file=hive.log
 
@@ -86,3 +86,12 @@ log4j.category.JPOX.Enhancer=ERROR,DRFA
 # Silence useless ZK logs
 log4j.logger.org.apache.zookeeper.server.NIOServerCnxn=WARN,DRFA
 log4j.logger.org.apache.zookeeper.ClientCnxnSocketNIO=WARN,DRFA
+
+#custom logging levels
+log4j.logger.org.apache.hadoop.hive.ql.parse.SemanticAnalyzer=INFO
+log4j.logger.org.apache.hadoop.hive.ql.Driver=INFO
+log4j.logger.org.apache.hadoop.hive.ql.exec.mr.ExecDriver=INFO
+log4j.logger.org.apache.hadoop.hive.ql.exec.mr.MapRedTask=INFO
+log4j.logger.org.apache.hadoop.hive.ql.exec.mr.MapredLocalTask=INFO
+log4j.logger.org.apache.hadoop.hive.ql.exec.Task=INFO
+log4j.logger.org.apache.hadoop.hive.ql.session.SessionState=INFO
diff --git a/common/src/test/resources/core-site.xml b/common/src/test/resources/core-site.xml
index ffe9747..ba53a17 100644
--- a/common/src/test/resources/core-site.xml
+++ b/common/src/test/resources/core-site.xml
@@ -24,4 +24,8 @@
     <name>fs.default.name</name>
     <value>core-site.xml</value>
   </property>
+    <property>
+    <name>mapred.reduce.tasks</name>
+    <value>1</value>
+  </property>
 </configuration>
diff --git a/contrib/pom.xml b/contrib/pom.xml
index 050c2d1..7e49b62 100644
--- a/contrib/pom.xml
+++ b/contrib/pom.xml
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
@@ -67,11 +67,21 @@
       <version>${junit.version}</version>
       <scope>test</scope>
     </dependency>
+    <!-- Following have only one version, which is cdh.hadoop.version -->
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-common</artifactId>
+      <version>${hadoop-23.version}</version>
+      <optional>true</optional>
+    </dependency>
   </dependencies>
 
   <profiles>
     <profile>
       <id>hadoop-1</id>
+      <activation>
+        <activeByDefault>false</activeByDefault>
+      </activation>
       <dependencies>
         <dependency>
           <groupId>org.apache.hadoop</groupId>
@@ -86,12 +96,6 @@
       <dependencies>
         <dependency>
           <groupId>org.apache.hadoop</groupId>
-          <artifactId>hadoop-common</artifactId>
-          <version>${hadoop-23.version}</version>
-          <optional>true</optional>
-        </dependency>
-        <dependency>
-          <groupId>org.apache.hadoop</groupId>
           <artifactId>hadoop-mapreduce-client-core</artifactId>
           <version>${hadoop-23.version}</version>
           <optional>true</optional>
diff --git a/data/conf/hive-site.xml b/data/conf/hive-site.xml
index ed3ee42..0ea8845 100644
--- a/data/conf/hive-site.xml
+++ b/data/conf/hive-site.xml
@@ -227,7 +227,6 @@
   <value>false</value>
 </property>
 
-
 <property>
   <name>hive.dummyparam.test.server.specific.config.override</name>
   <value>from.hive-site.xml</value>
@@ -240,7 +239,6 @@
   <description>Using dummy param to test server specific configuration</description>
 </property>
 
-
 <property>
   <name>hive.fetch.task.conversion</name>
   <value>minimal</value>
@@ -251,4 +249,10 @@
   <value>hive_admin_user</value>
 </property>
 
+<property>
+  <name>hive.security.authorization.task.factory</name>
+  <value>org.apache.hadoop.hive.ql.parse.authorization.HiveAuthorizationTaskFactoryImpl</value>
+  <description>The default SerDe hive will use for the rcfile format</description>
+</property>
+
 </configuration>
diff --git a/hbase-handler/pom.xml b/hbase-handler/pom.xml
index 1ffa501..6e6621d 100644
--- a/hbase-handler/pom.xml
+++ b/hbase-handler/pom.xml
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
@@ -67,6 +67,9 @@
   <profiles>
     <profile>
       <id>hadoop-1</id>
+      <activation>
+        <activeByDefault>false</activeByDefault>
+      </activation>
       <dependencies>
         <dependency>
           <groupId>org.apache.hadoop</groupId>
@@ -76,13 +79,13 @@
         </dependency>
         <dependency>
           <groupId>org.apache.hbase</groupId>
-          <artifactId>hbase-hadoop1-compat</artifactId>
+          <artifactId>hbase-hadoop-compat</artifactId>
           <version>${hbase.hadoop1.version}</version>
           <classifier>tests</classifier>
         </dependency>
         <dependency>
           <groupId>org.apache.hbase</groupId>
-          <artifactId>hbase-hadoop1-compat</artifactId>
+          <artifactId>hbase-hadoop2-compat</artifactId>
           <version>${hbase.hadoop1.version}</version>
         </dependency>
         <dependency>
@@ -165,7 +168,7 @@
         </dependency>
         <dependency>
           <groupId>org.apache.hbase</groupId>
-          <artifactId>hbase-hadoop2-compat</artifactId>
+          <artifactId>hbase-hadoop-compat</artifactId>
           <version>${hbase.hadoop2.version}</version>
           <classifier>tests</classifier>
         </dependency>
diff --git a/hcatalog/core/pom.xml b/hcatalog/core/pom.xml
index 307cfa7..8c008e0 100644
--- a/hcatalog/core/pom.xml
+++ b/hcatalog/core/pom.xml
@@ -1,3 +1,4 @@
+
 <!--
   Licensed to the Apache Software Foundation (ASF) under one
   or more contributor license agreements.  See the NOTICE file
@@ -25,7 +26,7 @@
   <parent>
     <groupId>org.apache.hive.hcatalog</groupId>
     <artifactId>hive-hcatalog</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
@@ -78,11 +79,62 @@
       <artifactId>jackson-mapper-asl</artifactId>
       <version>${jackson.version}</version>
     </dependency>
+      <dependency>
+      <groupId>com.sun.jersey</groupId>
+      <artifactId>jersey-servlet</artifactId>
+      <version>${jersey.version}</version>
+      <scope>test</scope>
+    </dependency>
+    <!-- Following have only one version, which is cdh.hadoop.version -->
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-annotations</artifactId>
+      <version>${hadoop-23.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-archives</artifactId>
+      <version>${hadoop-23.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-common</artifactId>
+      <version>${hadoop-23.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-common</artifactId>
+      <version>${hadoop-23.version}</version>
+      <classifier>tests</classifier>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-hdfs</artifactId>
+      <version>${hadoop-23.version}</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-hdfs</artifactId>
+      <version>${hadoop-23.version}</version>
+      <classifier>tests</classifier>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.pig</groupId>
+      <artifactId>pig</artifactId>
+      <version>${pig.version}</version>
+      <scope>test</scope>
+    </dependency>
   </dependencies>
 
   <profiles>
     <profile>
       <id>hadoop-1</id>
+      <activation>
+        <activeByDefault>false</activeByDefault>
+      </activation>
       <dependencies>
         <dependency>
           <groupId>org.apache.hadoop</groupId>
@@ -101,12 +153,6 @@
           <version>${hadoop-20S.version}</version>
           <scope>test</scope>
         </dependency>
-        <dependency>
-          <groupId>org.apache.pig</groupId>
-          <artifactId>pig</artifactId>
-          <version>${pig.version}</version>
-          <scope>test</scope>
-        </dependency>
       </dependencies>
     </profile>
    <profile>
@@ -114,21 +160,6 @@
       <dependencies>
         <dependency>
           <groupId>org.apache.hadoop</groupId>
-          <artifactId>hadoop-annotations</artifactId>
-          <version>${hadoop-23.version}</version>
-        </dependency>
-        <dependency>
-          <groupId>org.apache.hadoop</groupId>
-          <artifactId>hadoop-archives</artifactId>
-          <version>${hadoop-23.version}</version>
-        </dependency>
-        <dependency>
-          <groupId>org.apache.hadoop</groupId>
-          <artifactId>hadoop-common</artifactId>
-          <version>${hadoop-23.version}</version>
-        </dependency>
-        <dependency>
-          <groupId>org.apache.hadoop</groupId>
           <artifactId>hadoop-mapreduce-client-core</artifactId>
           <version>${hadoop-23.version}</version>
         </dependency>
@@ -139,26 +170,6 @@
         </dependency>
         <!-- test -->
         <dependency>
-          <groupId>com.sun.jersey</groupId>
-          <artifactId>jersey-servlet</artifactId>
-          <version>${jersey.version}</version>
-          <scope>test</scope>
-        </dependency>
-        <dependency>
-          <groupId>org.apache.hadoop</groupId>
-          <artifactId>hadoop-common</artifactId>
-          <version>${hadoop-23.version}</version>
-          <classifier>tests</classifier>
-          <scope>test</scope>
-        </dependency>
-        <dependency>
-          <groupId>org.apache.hadoop</groupId>
-          <artifactId>hadoop-hdfs</artifactId>
-          <version>${hadoop-23.version}</version>
-          <classifier>tests</classifier>
-          <scope>test</scope>
-        </dependency>
-        <dependency>
           <groupId>org.apache.hadoop</groupId>
           <artifactId>hadoop-mapreduce-client-hs</artifactId>
           <version>${hadoop-23.version}</version>
@@ -188,8 +199,13 @@
           <groupId>org.apache.pig</groupId>
           <artifactId>pig</artifactId>
           <version>${pig.version}</version>
-          <classifier>h2</classifier>
           <scope>test</scope>
+          <exclusions>
+            <exclusion>
+              <groupId>org.apache.hadoop</groupId>
+              <artifactId>hadoop-core</artifactId>
+            </exclusion>
+          </exclusions>
         </dependency>
       </dependencies>
     </profile>
diff --git a/hcatalog/core/src/main/java/org/apache/hive/hcatalog/mapreduce/MultiOutputFormat.java b/hcatalog/core/src/main/java/org/apache/hive/hcatalog/mapreduce/MultiOutputFormat.java
index 69a7345..c660e79 100644
--- a/hcatalog/core/src/main/java/org/apache/hive/hcatalog/mapreduce/MultiOutputFormat.java
+++ b/hcatalog/core/src/main/java/org/apache/hive/hcatalog/mapreduce/MultiOutputFormat.java
@@ -49,7 +49,6 @@
 import org.apache.hadoop.mapreduce.TaskAttemptContext;
 import org.apache.hadoop.mapreduce.TaskInputOutputContext;
 import org.apache.hadoop.util.ReflectionUtils;
-import org.apache.hive.hcatalog.common.HCatUtil;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -158,7 +157,7 @@
     configsToMerge.put(ShimLoader.getHadoopShims().getHCatShim().getPropertyName(
         HadoopShims.HCatHadoopShims.PropertyName.CACHE_FILES), COMMA_DELIM);
     String fileSep;
-    if (HCatUtil.isHadoop23()) {
+    if (org.apache.hadoop.mapred.MRVersion.isMR2()) {
       fileSep = ",";
     } else {
       fileSep = System.getProperty("path.separator");
@@ -361,7 +360,7 @@ private static JobConfigurer create(Job job) {
     public void addOutputFormat(String alias,
         Class<? extends OutputFormat> outputFormatClass,
         Class<?> keyClass, Class<?> valueClass) throws IOException {
-      Job copy = new Job(this.job.getConfiguration());
+      Job copy = Job.getInstance(this.job.getConfiguration());
       outputConfigs.put(alias, copy);
       copy.setOutputFormatClass(outputFormatClass);
       copy.setOutputKeyClass(keyClass);
diff --git a/hcatalog/core/src/test/java/org/apache/hive/hcatalog/mapreduce/TestMultiOutputFormat.java b/hcatalog/core/src/test/java/org/apache/hive/hcatalog/mapreduce/TestMultiOutputFormat.java
index d4271e5..237ed85 100644
--- a/hcatalog/core/src/test/java/org/apache/hive/hcatalog/mapreduce/TestMultiOutputFormat.java
+++ b/hcatalog/core/src/test/java/org/apache/hive/hcatalog/mapreduce/TestMultiOutputFormat.java
@@ -64,7 +64,8 @@
 
   private static final Logger LOG = LoggerFactory.getLogger(TestMultiOutputFormat.class);
   private static File workDir;
-  private static JobConf mrConf = null;
+
+  private Configuration mrConf = null;
   private static FileSystem fs = null;
   private static MiniMRCluster mrCluster = null;
 
@@ -79,8 +80,8 @@ public static void setup() throws IOException {
     System.setProperty("hadoop.log.dir", new File(workDir, "/logs").getAbsolutePath());
     // LocalJobRunner does not work with mapreduce OutputCommitter. So need
     // to use MiniMRCluster. MAPREDUCE-2350
-    mrConf = new JobConf(conf);
-    mrCluster = new MiniMRCluster(1, fs.getUri().toString(), 1, null, null, mrConf);
+    mrCluster = new MiniMRCluster(1, fs.getUri().toString(), 1, null, null,
+      new JobConf(conf));
   }
 
   private static void createWorkDir() throws IOException {
@@ -106,7 +107,8 @@ public static void tearDown() throws IOException {
    */
   @Test
   public void testMultiOutputFormatWithoutReduce() throws Throwable {
-    Job job = new Job(mrConf, "MultiOutNoReduce");
+    mrConf = mrCluster.createJobConf();
+    Job job = Job.getInstance(mrConf, "MultiOutNoReduce");
     job.setMapperClass(MultiOutWordIndexMapper.class);
     job.setJarByClass(this.getClass());
     job.setInputFormatClass(TextInputFormat.class);
@@ -173,7 +175,8 @@ public void testMultiOutputFormatWithoutReduce() throws Throwable {
    */
   @Test
   public void testMultiOutputFormatWithReduce() throws Throwable {
-    Job job = new Job(mrConf, "MultiOutWithReduce");
+    mrConf = mrCluster.createJobConf();
+    Job job = Job.getInstance(mrConf, "MultiOutWithReduce");
 
     job.setMapperClass(WordCountMapper.class);
     job.setReducerClass(MultiOutWordCountReducer.class);
diff --git a/hcatalog/hcatalog-pig-adapter/pom.xml b/hcatalog/hcatalog-pig-adapter/pom.xml
index d354f4d..193e01d 100644
--- a/hcatalog/hcatalog-pig-adapter/pom.xml
+++ b/hcatalog/hcatalog-pig-adapter/pom.xml
@@ -25,7 +25,7 @@
   <parent>
     <groupId>org.apache.hive.hcatalog</groupId>
     <artifactId>hive-hcatalog</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
@@ -72,6 +72,9 @@
   <profiles>
     <profile>
       <id>hadoop-1</id>
+      <activation>
+        <activeByDefault>false</activeByDefault>
+      </activation>
       <dependencies>
         <dependency>
           <groupId>org.apache.hadoop</groupId>
@@ -109,7 +112,12 @@
           <groupId>org.apache.pig</groupId>
           <artifactId>pig</artifactId>
           <version>${pig.version}</version>
-          <classifier>h2</classifier>
+          <exclusions>
+            <exclusion>
+              <groupId>org.apache.hadoop</groupId>
+              <artifactId>hadoop-core</artifactId>
+            </exclusion>
+          </exclusions>
         </dependency>
         <dependency>
           <!--this should be automatically brought in by Pig, it's not in Pig 0.12 due to a bug
diff --git a/hcatalog/hcatalog-pig-adapter/src/main/java/org/apache/hive/hcatalog/pig/HCatLoader.java b/hcatalog/hcatalog-pig-adapter/src/main/java/org/apache/hive/hcatalog/pig/HCatLoader.java
index c951847..3795cc1 100644
--- a/hcatalog/hcatalog-pig-adapter/src/main/java/org/apache/hive/hcatalog/pig/HCatLoader.java
+++ b/hcatalog/hcatalog-pig-adapter/src/main/java/org/apache/hive/hcatalog/pig/HCatLoader.java
@@ -123,7 +123,7 @@ public void setLocation(String location, Job job) throws IOException {
         job.getCredentials().addAll(crd);
       }
     } else {
-      Job clone = new Job(job.getConfiguration());
+      Job clone = Job.getInstance(job.getConfiguration());
       HCatInputFormat.setInput(job, dbName, tableName, getPartitionFilterString());
 
       InputJobInfo inputJobInfo = (InputJobInfo) HCatUtil.deserialize(
diff --git a/hcatalog/hcatalog-pig-adapter/src/main/java/org/apache/hive/hcatalog/pig/HCatStorer.java b/hcatalog/hcatalog-pig-adapter/src/main/java/org/apache/hive/hcatalog/pig/HCatStorer.java
index 75f133d..617eadf 100644
--- a/hcatalog/hcatalog-pig-adapter/src/main/java/org/apache/hive/hcatalog/pig/HCatStorer.java
+++ b/hcatalog/hcatalog-pig-adapter/src/main/java/org/apache/hive/hcatalog/pig/HCatStorer.java
@@ -163,7 +163,7 @@ public void setStoreLocation(String location, Job job) throws IOException {
         job.getCredentials().addAll(crd);
       }
     } else {
-      Job clone = new Job(job.getConfiguration());
+      Job clone = Job.getInstance(job.getConfiguration());
       OutputJobInfo outputJobInfo;
       if (userStr.length == 2) {
         outputJobInfo = OutputJobInfo.create(userStr[0], userStr[1], partitions);
diff --git a/hcatalog/pom.xml b/hcatalog/pom.xml
index f4d0ee3..6460e73 100644
--- a/hcatalog/pom.xml
+++ b/hcatalog/pom.xml
@@ -24,7 +24,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
@@ -58,6 +58,9 @@
   <profiles>
     <profile>
       <id>hadoop-1</id>
+      <activation>
+        <activeByDefault>false</activeByDefault>
+      </activation>
       <dependencies>
         <dependency>
           <groupId>org.apache.hadoop</groupId>
@@ -92,8 +95,13 @@
           <groupId>org.apache.pig</groupId>
           <artifactId>pig</artifactId>
           <version>${pig.version}</version>
-          <classifier>h2</classifier>
           <scope>test</scope>
+          <exclusions>
+            <exclusion>
+              <groupId>org.apache.hadoop</groupId>
+              <artifactId>hadoop-core</artifactId>
+            </exclusion>
+          </exclusions>
         </dependency>
       </dependencies>
     </profile>
diff --git a/hcatalog/server-extensions/pom.xml b/hcatalog/server-extensions/pom.xml
index 07b2976..46ee5f3 100644
--- a/hcatalog/server-extensions/pom.xml
+++ b/hcatalog/server-extensions/pom.xml
@@ -25,7 +25,7 @@
   <parent>
     <groupId>org.apache.hive.hcatalog</groupId>
     <artifactId>hive-hcatalog</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
@@ -94,12 +94,21 @@
       <artifactId>pig</artifactId>
       <version>${pig.version}</version>
       <scope>test</scope>
+      <exclusions>
+        <exclusion>
+          <groupId>org.apache.hadoop</groupId>
+          <artifactId>hadoop-core</artifactId>
+        </exclusion>
+      </exclusions>
     </dependency>
   </dependencies>
 
   <profiles>
     <profile>
       <id>hadoop-1</id>
+      <activation>
+        <activeByDefault>false</activeByDefault>
+      </activation>
       <dependencies>
         <dependency>
           <groupId>org.apache.hadoop</groupId>
diff --git a/hcatalog/streaming/pom.xml b/hcatalog/streaming/pom.xml
index df14c90..f02f84b 100644
--- a/hcatalog/streaming/pom.xml
+++ b/hcatalog/streaming/pom.xml
@@ -20,7 +20,7 @@
   <parent>
     <groupId>org.apache.hive.hcatalog</groupId>
     <artifactId>hive-hcatalog</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
diff --git a/hcatalog/webhcat/java-client/pom.xml b/hcatalog/webhcat/java-client/pom.xml
index 2c5678e..ef6dc5a 100644
--- a/hcatalog/webhcat/java-client/pom.xml
+++ b/hcatalog/webhcat/java-client/pom.xml
@@ -25,7 +25,7 @@
   <parent>
     <groupId>org.apache.hive.hcatalog</groupId>
     <artifactId>hive-hcatalog</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../../pom.xml</relativePath>
   </parent>
 
@@ -65,6 +65,9 @@
   <profiles>
     <profile>
       <id>hadoop-1</id>
+      <activation>
+        <activeByDefault>false</activeByDefault>
+      </activation>
       <dependencies>
         <dependency>
           <groupId>org.apache.hadoop</groupId>
diff --git a/hcatalog/webhcat/svr/pom.xml b/hcatalog/webhcat/svr/pom.xml
index d8eb111..38549d5 100644
--- a/hcatalog/webhcat/svr/pom.xml
+++ b/hcatalog/webhcat/svr/pom.xml
@@ -25,7 +25,7 @@
   <parent>
     <groupId>org.apache.hive.hcatalog</groupId>
     <artifactId>hive-hcatalog</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../../pom.xml</relativePath>
   </parent>
 
@@ -109,12 +109,31 @@
       <version>${junit.version}</version>
       <scope>test</scope>
     </dependency>
+    <!-- Following have only one version, which is cdh.hadoop.version -->
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-auth</artifactId>
+      <version>${hadoop-23.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-common</artifactId>
+      <version>${hadoop-23.version}</version>
+    </dependency>
+      <dependency>
+      <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-hdfs</artifactId>
+      <version>${hadoop-23.version}</version>
+      </dependency>
   </dependencies>
 
 
   <profiles>
     <profile>
       <id>hadoop-1</id>
+      <activation>
+        <activeByDefault>false</activeByDefault>
+      </activation>
       <dependencies>
         <dependency>
           <groupId>org.apache.hadoop</groupId>
@@ -128,21 +147,6 @@
       <dependencies>
         <dependency>
           <groupId>org.apache.hadoop</groupId>
-          <artifactId>hadoop-auth</artifactId>
-          <version>${hadoop-23.version}</version>
-        </dependency>
-        <dependency>
-          <groupId>org.apache.hadoop</groupId>
-          <artifactId>hadoop-common</artifactId>
-          <version>${hadoop-23.version}</version>
-        </dependency>
-        <dependency>
-          <groupId>org.apache.hadoop</groupId>
-          <artifactId>hadoop-hdfs</artifactId>
-          <version>${hadoop-23.version}</version>
-        </dependency>
-        <dependency>
-          <groupId>org.apache.hadoop</groupId>
           <artifactId>hadoop-mapreduce-client-core</artifactId>
           <version>${hadoop-23.version}</version>
         </dependency>
diff --git a/hwi/pom.xml b/hwi/pom.xml
index 5e51632..50d92e0 100644
--- a/hwi/pom.xml
+++ b/hwi/pom.xml
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
@@ -86,11 +86,20 @@
       <version>${junit.version}</version>
       <scope>test</scope>
     </dependency>
+    <dependency>
+       <groupId>org.apache.hadoop</groupId>
+       <artifactId>hadoop-common</artifactId>
+       <version>${hadoop-23.version}</version>
+       <optional>true</optional>
+     </dependency>
   </dependencies>
 
   <profiles>
     <profile>
       <id>hadoop-1</id>
+      <activation>
+        <activeByDefault>false</activeByDefault>
+      </activation>
       <dependencies>
     <!-- dependencies are always listed in sorted order by groupId, artifectId -->
         <dependency>
diff --git a/itests/custom-serde/pom.xml b/itests/custom-serde/pom.xml
index 6c75278..23b1a8b 100644
--- a/itests/custom-serde/pom.xml
+++ b/itests/custom-serde/pom.xml
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive-it</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
@@ -44,6 +44,9 @@
   <profiles>
     <profile>
       <id>hadoop-1</id>
+      <activation>
+        <activeByDefault>false</activeByDefault>
+      </activation>
       <dependencies>
         <dependency>
           <groupId>org.apache.hadoop</groupId>
diff --git a/itests/hcatalog-unit/pom.xml b/itests/hcatalog-unit/pom.xml
index c728049..1c68690 100644
--- a/itests/hcatalog-unit/pom.xml
+++ b/itests/hcatalog-unit/pom.xml
@@ -25,7 +25,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive-it</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
@@ -114,11 +114,48 @@
       <version>${junit.version}</version>
       <scope>test</scope>
     </dependency>
+    <!-- Following have only one version, which is cdh.hadoop.version -->
+    <dependency>
+      <groupId>org.apache.pig</groupId>
+      <artifactId>pig</artifactId>
+      <version>${pig.version}</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+       <!--this should be automatically brought in by Pig, it's not in Pig 0.12 due to a bug
+          in Pig which requires it This is fixed in Pig's pom file in ASF trunk (pig 13)-->
+      <groupId>joda-time</groupId>
+      <artifactId>joda-time</artifactId>
+      <version>2.2</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-annotations</artifactId>
+      <version>${hadoop-23.version}</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-archives</artifactId>
+      <version>${hadoop-23.version}</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-common</artifactId>
+      <version>${hadoop-23.version}</version>
+      <classifier>tests</classifier>
+      <scope>test</scope>
+    </dependency>
   </dependencies>
 
   <profiles>
     <profile>
       <id>hadoop-1</id>
+      <activation>
+        <activeByDefault>false</activeByDefault>
+      </activation>
       <dependencies>
         <!-- test -->
         <dependency>
@@ -173,13 +210,13 @@
         </dependency>
         <dependency>
           <groupId>org.apache.hbase</groupId>
-          <artifactId>hbase-hadoop1-compat</artifactId>
+          <artifactId>hbase-hadoop2-compat</artifactId>
           <version>${hbase.hadoop1.version}</version>
           <scope>test</scope>
         </dependency>
         <dependency>
           <groupId>org.apache.hbase</groupId>
-          <artifactId>hbase-hadoop1-compat</artifactId>
+          <artifactId>hbase-hadoop2-compat</artifactId>
           <version>${hbase.hadoop1.version}</version>
           <scope>test</scope>
           <classifier>tests</classifier>
@@ -197,20 +234,6 @@
           <classifier>tests</classifier>
           <scope>test</scope>
         </dependency>
-        <dependency>
-          <groupId>org.apache.pig</groupId>
-          <artifactId>pig</artifactId>
-          <version>${pig.version}</version>
-          <scope>test</scope>
-        </dependency>
-        <dependency>
-          <!--this should be automatically brought in by Pig, it's not in Pig 0.12 due to a bug
-              in Pig which requires it This is fixed in Pig's pom file in ASF trunk (pig 13)-->
-          <groupId>joda-time</groupId>
-          <artifactId>joda-time</artifactId>
-          <version>2.2</version>
-          <scope>test</scope>
-        </dependency>
       </dependencies>
     </profile>
    <profile>
@@ -219,31 +242,6 @@
         <!-- test -->
         <dependency>
           <groupId>org.apache.hadoop</groupId>
-          <artifactId>hadoop-annotations</artifactId>
-          <version>${hadoop-23.version}</version>
-          <scope>test</scope>
-        </dependency>
-        <dependency>
-          <groupId>org.apache.hadoop</groupId>
-          <artifactId>hadoop-archives</artifactId>
-          <version>${hadoop-23.version}</version>
-          <scope>test</scope>
-        </dependency>
-        <dependency>
-          <groupId>org.apache.hadoop</groupId>
-          <artifactId>hadoop-common</artifactId>
-          <version>${hadoop-23.version}</version>
-          <scope>test</scope>
-        </dependency>
-        <dependency>
-          <groupId>org.apache.hadoop</groupId>
-          <artifactId>hadoop-common</artifactId>
-          <version>${hadoop-23.version}</version>
-          <classifier>tests</classifier>
-          <scope>test</scope>
-        </dependency>
-        <dependency>
-          <groupId>org.apache.hadoop</groupId>
           <artifactId>hadoop-hdfs</artifactId>
           <version>${hadoop-23.version}</version>
           <scope>test</scope>
@@ -348,8 +346,13 @@
           <groupId>org.apache.pig</groupId>
           <artifactId>pig</artifactId>
           <version>${pig.version}</version>
-          <classifier>h2</classifier>
           <scope>test</scope>
+          <exclusions>
+            <exclusion>
+              <groupId>org.apache.hadoop</groupId>
+              <artifactId>hadoop-core</artifactId>
+            </exclusion>
+          </exclusions>
         </dependency>
         <dependency>
           <!--this should be automatically brought in by Pig, it's not in Pig 0.12 due to a bug
diff --git a/itests/hive-jmh/pom.xml b/itests/hive-jmh/pom.xml
index 7b30242..d8f9fb6 100644
--- a/itests/hive-jmh/pom.xml
+++ b/itests/hive-jmh/pom.xml
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive-it</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
diff --git a/itests/hive-minikdc/pom.xml b/itests/hive-minikdc/pom.xml
index f908164..af87e0f 100644
--- a/itests/hive-minikdc/pom.xml
+++ b/itests/hive-minikdc/pom.xml
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive-it</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
@@ -175,13 +175,13 @@
         </dependency>
         <dependency>
           <groupId>org.apache.hbase</groupId>
-          <artifactId>hbase-hadoop1-compat</artifactId>
+          <artifactId>hbase-hadoop2-compat</artifactId>
           <version>${hbase.hadoop1.version}</version>
           <scope>test</scope>
         </dependency>
         <dependency>
           <groupId>org.apache.hbase</groupId>
-          <artifactId>hbase-hadoop1-compat</artifactId>
+          <artifactId>hbase-hadoop2-compat</artifactId>
           <version>${hbase.hadoop1.version}</version>
           <scope>test</scope>
           <classifier>tests</classifier>
diff --git a/itests/hive-unit-hadoop2/pom.xml b/itests/hive-unit-hadoop2/pom.xml
index 474963e..2534359 100644
--- a/itests/hive-unit-hadoop2/pom.xml
+++ b/itests/hive-unit-hadoop2/pom.xml
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive-it</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
diff --git a/itests/hive-unit/pom.xml b/itests/hive-unit/pom.xml
index a28dadc..5684349 100644
--- a/itests/hive-unit/pom.xml
+++ b/itests/hive-unit/pom.xml
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive-it</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
@@ -133,6 +133,9 @@
   <profiles>
     <profile>
       <id>hadoop-1</id>
+      <activation>
+        <activeByDefault>false</activeByDefault>
+      </activation>
       <dependencies>
         <dependency>
           <groupId>org.apache.hadoop</groupId>
@@ -173,13 +176,13 @@
         </dependency>
         <dependency>
           <groupId>org.apache.hbase</groupId>
-          <artifactId>hbase-hadoop1-compat</artifactId>
+          <artifactId>hbase-hadoop2-compat</artifactId>
           <version>${hbase.hadoop1.version}</version>
           <scope>test</scope>
         </dependency>
         <dependency>
           <groupId>org.apache.hbase</groupId>
-          <artifactId>hbase-hadoop1-compat</artifactId>
+          <artifactId>hbase-hadoop2-compat</artifactId>
           <version>${hbase.hadoop1.version}</version>
           <scope>test</scope>
           <classifier>tests</classifier>
@@ -208,6 +211,12 @@
           <artifactId>jersey-servlet</artifactId>
           <scope>test</scope>
         </dependency>
+        <dependency>
+          <groupId>org.apache.hbase</groupId>
+          <artifactId>hbase-server</artifactId>
+          <version>${hbase.hadoop1.version}</version>
+          <scope>test</scope>
+        </dependency>
       </dependencies>
     </profile>
     <profile>
@@ -241,7 +250,7 @@
         <dependency>
           <groupId>org.apache.hbase</groupId>
           <artifactId>hbase-server</artifactId>
-          <version>${hbase.hadoop2.version}</version>
+          <version>${hbase.hadoop1.version}</version>
           <scope>test</scope>
         </dependency>
         <dependency>
diff --git a/itests/hive-unit/src/test/java/org/apache/hive/jdbc/TestSSL.java b/itests/hive-unit/src/test/java/org/apache/hive/jdbc/TestSSL.java
index 30c4e99..86c9f58 100644
--- a/itests/hive-unit/src/test/java/org/apache/hive/jdbc/TestSSL.java
+++ b/itests/hive-unit/src/test/java/org/apache/hive/jdbc/TestSSL.java
@@ -412,7 +412,29 @@ private void setHttpConfOverlay(Map<String, String> confOverlay) {
 
   private void setBinaryConfOverlay(Map<String, String> confOverlay) {
     confOverlay.put(ConfVars.HIVE_SERVER2_TRANSPORT_MODE.varname, HS2_BINARY_MODE);
-    confOverlay.put(ConfVars.HIVE_SERVER2_AUTHENTICATION.varname,  HS2_BINARY_AUTH_MODE);
+    confOverlay.put(ConfVars.HIVE_SERVER2_AUTHENTICATION.varname, HS2_BINARY_AUTH_MODE);
     confOverlay.put(ConfVars.HIVE_SERVER2_ENABLE_DOAS.varname, "true");
   }
+
+  /***
+   * Test SSL client connection to SSL server
+   * @throws Exception
+   */
+  @Test
+  public void testSSLDeprecatConfig() throws Exception {
+    // Start HS2 with SSL using old config
+    miniHS2.setConfProperty("hive.server2.enable.SSL", "true");
+    miniHS2.setConfProperty(ConfVars.HIVE_SERVER2_SSL_KEYSTORE_PATH.varname,
+        dataFileDir + File.separator +  KEY_STORE_NAME);
+    miniHS2.setConfProperty(ConfVars.HIVE_SERVER2_SSL_KEYSTORE_PASSWORD.varname,
+        KEY_STORE_PASSWORD);
+    miniHS2.start(new HashMap<String, String>());
+
+    // make SSL connection
+    hs2Conn = DriverManager.getConnection(miniHS2.getJdbcURL() + ";ssl=true;sslTrustStore=" +
+        dataFileDir + File.separator + TRUST_STORE_NAME + ";trustStorePassword=" +
+        KEY_STORE_PASSWORD, System.getProperty("user.name"), "bar");
+
+    hs2Conn.close();
+  }
 }
diff --git a/itests/pom.xml b/itests/pom.xml
index c994c1a..448a059 100644
--- a/itests/pom.xml
+++ b/itests/pom.xml
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
@@ -91,7 +91,7 @@
                     mv $BASE_DIR/spark-${spark.version}-bin-hadoop2-without-hive $BASE_DIR/$finalName
                   }
                   mkdir -p $DOWNLOAD_DIR
-                  download "http://d3jw87u4immizc.cloudfront.net/spark-tarball/spark-${spark.version}-bin-hadoop2-without-hive.tgz" "spark"
+                  download "http://d3jw87u4immizc.cloudfront.net/spark-tarball/spark-1.2.0-bin-hadoop2-without-hive.tgz" "spark"
                   cp -f $HIVE_ROOT/data/conf/spark/log4j.properties $BASE_DIR/spark/conf/
                 </echo>
               </target>
diff --git a/itests/qtest-spark/pom.xml b/itests/qtest-spark/pom.xml
index ed39e2a..2247145 100644
--- a/itests/qtest-spark/pom.xml
+++ b/itests/qtest-spark/pom.xml
@@ -20,7 +20,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive-it</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
diff --git a/itests/qtest/pom.xml b/itests/qtest/pom.xml
index 1652af8..d318610 100644
--- a/itests/qtest/pom.xml
+++ b/itests/qtest/pom.xml
@@ -20,7 +20,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive-it</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
@@ -119,13 +119,29 @@
     </profile>
     <profile>
       <id>hadoop-1</id>
+      <activation>
+        <activeByDefault>false</activeByDefault>
+      </activation>
       <properties>
         <active.hadoop.version>${hadoop-20S.version}</active.hadoop.version>
-        <test.dfs.mkdir>-mkdir</test.dfs.mkdir>
+        <test.dfs.mkdir>-mkdir -p</test.dfs.mkdir>
       </properties>
       <dependencies>
         <dependency>
           <groupId>org.apache.hadoop</groupId>
+          <artifactId>hadoop-common</artifactId>
+          <version>${hadoop-23.version}</version>
+          <scope>test</scope>
+        </dependency>
+        <dependency>
+          <groupId>org.apache.hadoop</groupId>
+          <artifactId>hadoop-common</artifactId>
+          <version>${hadoop-23.version}</version>
+          <scope>test</scope>
+          <classifier>tests</classifier>
+        </dependency>
+        <dependency>
+          <groupId>org.apache.hadoop</groupId>
           <artifactId>hadoop-core</artifactId>
           <version>${hadoop-20S.version}</version>
           <scope>test</scope>
@@ -170,13 +186,13 @@
         </dependency>
         <dependency>
           <groupId>org.apache.hbase</groupId>
-          <artifactId>hbase-hadoop1-compat</artifactId>
+          <artifactId>hbase-hadoop2-compat</artifactId>
           <version>${hbase.hadoop1.version}</version>
           <scope>test</scope>
         </dependency>
         <dependency>
           <groupId>org.apache.hbase</groupId>
-          <artifactId>hbase-hadoop1-compat</artifactId>
+          <artifactId>hbase-hadoop2-compat</artifactId>
           <version>${hbase.hadoop1.version}</version>
           <scope>test</scope>
           <classifier>tests</classifier>
@@ -211,6 +227,11 @@
         </dependency>
         <dependency>
           <groupId>org.apache.hadoop</groupId>
+          <artifactId>hadoop-mapreduce-client-jobclient</artifactId>
+          <version>${hadoop-23.version}</version>
+        </dependency>
+        <dependency>
+          <groupId>org.apache.hadoop</groupId>
           <artifactId>hadoop-archives</artifactId>
           <version>${hadoop-23.version}</version>
           <scope>test</scope>
diff --git a/itests/test-serde/pom.xml b/itests/test-serde/pom.xml
index 7a4d164..0c730b0 100644
--- a/itests/test-serde/pom.xml
+++ b/itests/test-serde/pom.xml
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive-it</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
@@ -45,6 +45,9 @@
   <profiles>
     <profile>
       <id>hadoop-1</id>
+      <activation>
+        <activeByDefault>false</activeByDefault>
+      </activation>
       <dependencies>
         <dependency>
           <groupId>org.apache.hadoop</groupId>
diff --git a/itests/util/pom.xml b/itests/util/pom.xml
index c370ffe..ce8bc7c 100644
--- a/itests/util/pom.xml
+++ b/itests/util/pom.xml
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive-it</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
@@ -97,6 +97,9 @@
   <profiles>
     <profile>
       <id>hadoop-1</id>
+      <activation>
+        <activeByDefault>false</activeByDefault>
+      </activation>
       <dependencies>
         <dependency>
           <groupId>org.apache.hadoop</groupId>
diff --git a/itests/util/src/main/java/org/apache/hadoop/hive/ql/hooks/VerifyOverriddenConfigsHook.java b/itests/util/src/main/java/org/apache/hadoop/hive/ql/hooks/VerifyOverriddenConfigsHook.java
index 57baef0..aa01baf 100644
--- a/itests/util/src/main/java/org/apache/hadoop/hive/ql/hooks/VerifyOverriddenConfigsHook.java
+++ b/itests/util/src/main/java/org/apache/hadoop/hive/ql/hooks/VerifyOverriddenConfigsHook.java
@@ -38,7 +38,7 @@
   // a config variable not in the default List of config variables, and a config variable in the
   // default list of config variables, but which has not been overridden
   private static String[] keysArray =
-    {"mapred.job.tracker", "hive.exec.post.hooks", "some.hive.config.doesnt.exit",
+    {"hive.exec.post.hooks", "hive.config.doesnt.exit",
      "hive.exec.mode.local.auto"};
   private static List<String> keysList = Arrays.asList(keysArray);
 
diff --git a/jdbc/pom.xml b/jdbc/pom.xml
index fa32607..655be72 100644
--- a/jdbc/pom.xml
+++ b/jdbc/pom.xml
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
@@ -108,6 +108,9 @@
   <profiles>
     <profile>
       <id>hadoop-1</id>
+      <activation>
+        <activeByDefault>false</activeByDefault>
+      </activation>
       <dependencies>
         <dependency>
           <groupId>org.apache.hadoop</groupId>
diff --git a/metastore/pom.xml b/metastore/pom.xml
index 7b804d1..ea01283 100644
--- a/metastore/pom.xml
+++ b/metastore/pom.xml
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
@@ -134,11 +134,21 @@
       <version>${mockito-all.version}</version>
       <scope>test</scope>
     </dependency>
+    <!-- Following have only one version, which is cdh.hadoop.version -->
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-common</artifactId>
+      <version>${hadoop-23.version}</version>
+      <optional>true</optional>
+    </dependency>
   </dependencies>
 
   <profiles>
     <profile>
       <id>hadoop-1</id>
+      <activation>
+        <activeByDefault>false</activeByDefault>
+      </activation>
       <dependencies>
         <dependency>
           <groupId>org.apache.hadoop</groupId>
@@ -153,12 +163,6 @@
       <dependencies>
         <dependency>
           <groupId>org.apache.hadoop</groupId>
-          <artifactId>hadoop-common</artifactId>
-          <version>${hadoop-23.version}</version>
-          <optional>true</optional>
-        </dependency>
-        <dependency>
-          <groupId>org.apache.hadoop</groupId>
           <artifactId>hadoop-mapreduce-client-core</artifactId>
           <version>${hadoop-23.version}</version>
           <optional>true</optional>
diff --git a/odbc/pom.xml b/odbc/pom.xml
index 86eab99..b1cbb08 100644
--- a/odbc/pom.xml
+++ b/odbc/pom.xml
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
diff --git a/packaging/pom.xml b/packaging/pom.xml
index 717dbc3..4b6ada7 100644
--- a/packaging/pom.xml
+++ b/packaging/pom.xml
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
@@ -117,6 +117,16 @@
   <dependencies>
     <!-- dependencies are always listed in sorted order by groupId, artifectId -->
     <dependency>
+      <groupId>org.apache.hive.shims</groupId>
+      <artifactId>hive-shims-common</artifactId>
+      <version>${project.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.hive.shims</groupId>
+      <artifactId>hive-shims-0.23</artifactId>
+      <version>${project.version}</version>
+    </dependency>
+    <dependency>
       <groupId>org.apache.hive</groupId>
       <artifactId>hive-common</artifactId>
       <version>${project.version}</version>
diff --git a/pom.xml b/pom.xml
index 14124a4..d81e436 100644
--- a/pom.xml
+++ b/pom.xml
@@ -15,13 +15,14 @@
 <project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
   <modelVersion>4.0.0</modelVersion>
   <parent>
-    <groupId>org.apache</groupId>
-    <artifactId>apache</artifactId>
-    <version>14</version>
+    <groupId>com.cloudera.cdh</groupId>
+    <artifactId>cdh-root</artifactId>
+    <version>5.4.0-SNAPSHOT</version>
   </parent>
+
   <groupId>org.apache.hive</groupId>
   <artifactId>hive</artifactId>
-  <version>1.1.0</version>
+  <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
   <packaging>pom</packaging>
 
   <name>Hive</name>
@@ -53,6 +54,7 @@
   </modules>
 
   <properties>
+    <cdh.hive.version>1.1.0-cdh5.4.0-SNAPSHOT</cdh.hive.version>
     <hive.version.shortname>1.1.0</hive.version.shortname>
 
     <!-- Build Properties -->
@@ -98,12 +100,14 @@
     <activemq.version>5.5.0</activemq.version>
     <ant.version>1.9.1</ant.version>
     <antlr.version>3.4</antlr.version>
-    <avro.version>1.7.5</avro.version>
+    <avro.version>${cdh.avro.version}</avro.version>
     <bonecp.version>0.8.0.RELEASE</bonecp.version>
     <calcite.version>1.0.0-incubating</calcite.version>
     <datanucleus-api-jdo.version>3.2.6</datanucleus-api-jdo.version>
     <datanucleus-core.version>3.2.10</datanucleus-core.version>
     <datanucleus-rdbms.version>3.2.9</datanucleus-rdbms.version>
+    <checkstyle.version>5.0</checkstyle.version>
+    <findbugs.version>1.3.9</findbugs.version>
     <commons-cli.version>1.2</commons-cli.version>
     <commons-codec.version>1.4</commons-codec.version>
     <commons-compress.version>1.4.1</commons-compress.version>
@@ -118,14 +122,14 @@
     <derby.version>10.11.1.1</derby.version>
     <guava.version>14.0.1</guava.version>
     <groovy.version>2.1.6</groovy.version>
-    <hadoop-20S.version>1.2.1</hadoop-20S.version>
-    <hadoop-23.version>2.6.0</hadoop-23.version>
+    <hadoop-20S.version>${cdh.mr1.version}</hadoop-20S.version>
+    <hadoop-23.version>${cdh.hadoop.version}</hadoop-23.version>
     <hadoop.bin.path>${basedir}/${hive.path.to.root}/testutils/hadoop</hadoop.bin.path>
-    <hbase.hadoop1.version>0.98.3-hadoop1</hbase.hadoop1.version>
-    <hbase.hadoop2.version>0.98.3-hadoop2</hbase.hadoop2.version>
+    <hbase.hadoop1.version>${cdh.hbase.version}</hbase.hadoop1.version>
+    <hbase.hadoop2.version>${cdh.hbase.version}</hbase.hadoop2.version>
     <!-- httpcomponents are not always in version sync -->
-    <httpcomponents.client.version>4.2.5</httpcomponents.client.version>
-    <httpcomponents.core.version>4.2.5</httpcomponents.core.version>
+    <httpcomponents.client.version>${cdh.httpcomponents.version}</httpcomponents.client.version>
+    <httpcomponents.core.version>${cdh.httpcomponents.core.version}</httpcomponents.core.version>
     <jackson.version>1.9.2</jackson.version>
     <javaewah.version>0.3.2</javaewah.version>
     <javolution.version>5.5.1</javolution.version>
@@ -146,15 +150,15 @@
     <mockito-all.version>1.9.5</mockito-all.version>
     <mina.version>2.0.0-M5</mina.version>
     <netty.version>4.0.23.Final</netty.version>
-    <parquet.version>1.6.0rc3</parquet.version>
-    <pig.version>0.12.0</pig.version>
-    <protobuf.version>2.5.0</protobuf.version>
+    <parquet.version>${cdh.parquet.version}</parquet.version>
+    <pig.version>${cdh.pig.version}</pig.version>
+    <protobuf.version>${cdh.protobuf.version}</protobuf.version>
     <stax.version>1.0.1</stax.version>
-    <slf4j.version>1.7.5</slf4j.version>
+    <slf4j.version>${cdh.slf4j.version}</slf4j.version>
     <ST4.version>4.0.4</ST4.version>
     <tez.version>0.5.2</tez.version>
     <super-csv.version>2.2.0</super-csv.version>
-    <spark.version>1.2.0</spark.version>
+    <spark.version>${cdh.spark.version}</spark.version>
     <scala.binary.version>2.10</scala.binary.version>
     <scala.version>2.10.4</scala.version>
     <tempus-fugit.version>1.1</tempus-fugit.version>
@@ -162,7 +166,7 @@
     <wadl-resourcedoc-doclet.version>1.4</wadl-resourcedoc-doclet.version>
     <velocity.version>1.5</velocity.version>
     <xerces.version>2.9.1</xerces.version>
-    <zookeeper.version>3.4.6</zookeeper.version>
+    <zookeeper.version>${cdh.zookeeper.version}</zookeeper.version>
     <jpam.version>1.1</jpam.version>
     <felix.version>2.4.0</felix.version>
     <curator.version>2.6.0</curator.version>
@@ -172,6 +176,22 @@
   <repositories>
    <!-- This needs to be removed before checking in-->
     <repository>
+      <id>cdh.releases.repo</id>
+      <url>https://repository.cloudera.com/content/groups/cdh-releases-rcs</url>
+      <name>CDH Releases Repository</name>
+      <snapshots>
+        <enabled>false</enabled>
+      </snapshots>
+    </repository>
+    <repository>
+      <id>cdh.snapshots.repo</id>
+      <url>https://repository.cloudera.com/content/repositories/snapshots</url>
+      <name>CDH Snapshots Repository</name>
+      <snapshots>
+        <enabled>true</enabled>
+      </snapshots>
+    </repository>
+    <repository>
       <id>datanucleus</id>
       <name>datanucleus maven repository</name>
       <url>http://www.datanucleus.org/downloads/maven2</url>
@@ -214,6 +234,14 @@
          <enabled>false</enabled>
        </snapshots>
     </repository>
+    <repository>
+      <id>apache.snapshots</id>
+      <name>Apache Snapshot Repository</name>
+      <url>http://repository.apache.org/snapshots</url>
+      <releases>
+        <enabled>false</enabled>
+      </releases>
+    </repository>
   </repositories>
 
   <!-- Hadoop dependency management is done at the bottom under profiles -->
@@ -834,7 +862,7 @@
             <LANG>en_US.UTF-8</LANG>
             <HADOOP_CLASSPATH>${test.tmp.dir}/conf:${basedir}/${hive.path.to.root}/conf</HADOOP_CLASSPATH>
             <HIVE_HADOOP_TEST_CLASSPATH>${test.hive.hadoop.classpath}</HIVE_HADOOP_TEST_CLASSPATH>
-            <SPARK_SUBMIT_CLASSPATH>${spark.home}/lib/spark-assembly-${spark.version}-hadoop2.4.0.jar:${test.hive.hadoop.classpath}</SPARK_SUBMIT_CLASSPATH>
+            <SPARK_SUBMIT_CLASSPATH>${spark.home}/lib/spark-assembly-1.2.0-hadoop2.4.0.jar:${test.hive.hadoop.classpath}</SPARK_SUBMIT_CLASSPATH>
             <SPARK_OSX_TEST_OPTS>-Dorg.xerial.snappy.tempdir=/tmp -Dorg.xerial.snappy.lib.name=libsnappyjava.jnilib</SPARK_OSX_TEST_OPTS>
             <PATH>${env.PATH}${test.extra.path}</PATH>
           </environmentVariables>
@@ -1057,7 +1085,7 @@
           </dependency>
           <dependency>
             <groupId>org.apache.hbase</groupId>
-            <artifactId>hbase-hadoop1-compat</artifactId>
+            <artifactId>hbase-hadoop2-compat</artifactId>
             <version>${hbase.hadoop1.version}</version>
           </dependency>
           <dependency>
@@ -1070,6 +1098,14 @@
     </profile>
     <profile>
       <id>hadoop-2</id>
+      <!-- Force all deps to include MRVersion class, see CDH-22186 -->
+      <dependencies>
+        <dependency>
+          <groupId>org.apache.hadoop</groupId>
+          <artifactId>hadoop-mapreduce-client-common</artifactId>
+          <version>${hadoop-23.version}</version>
+        </dependency>
+      </dependencies>
       <dependencyManagement>
         <dependencies>
           <dependency>
@@ -1106,21 +1142,88 @@
             <groupId>org.apache.hbase</groupId>
             <artifactId>hbase-common</artifactId>
             <version>${hbase.hadoop2.version}</version>
+            <!-- hbase depends on hadoop-core which pollutes our classpath -->
+            <exclusions>
+              <exclusion>
+                <groupId>org.apache.hadoop</groupId>
+                <artifactId>hadoop-core</artifactId>
+              </exclusion>
+            </exclusions>
+          </dependency>
+          <dependency>
+            <groupId>org.apache.hbase</groupId>
+            <artifactId>hbase-common</artifactId>
+            <version>${hbase.hadoop2.version}</version>
+            <type>test-jar</type>
+            <!-- hbase depends on hadoop-core which pollutes our classpath -->
+            <exclusions>
+              <exclusion>
+                <groupId>org.apache.hadoop</groupId>
+                <artifactId>hadoop-core</artifactId>
+              </exclusion>
+            </exclusions>
           </dependency>
           <dependency>
             <groupId>org.apache.hbase</groupId>
             <artifactId>hbase-hadoop-compat</artifactId>
             <version>${hbase.hadoop2.version}</version>
+            <!-- hbase depends on hadoop-core which pollutes our classpath -->
+            <exclusions>
+              <exclusion>
+                <groupId>org.apache.hadoop</groupId>
+                <artifactId>hadoop-core</artifactId>
+              </exclusion>
+            </exclusions>
+          </dependency>
+          <dependency>
+            <groupId>org.apache.hbase</groupId>
+            <artifactId>hbase-hadoop2-compat</artifactId>
+            <version>${hbase.hadoop2.version}</version>
+            <!-- hbase depends on hadoop-core which pollutes our classpath -->
+            <exclusions>
+              <exclusion>
+                <groupId>org.apache.hadoop</groupId>
+                <artifactId>hadoop-core</artifactId>
+              </exclusion>
+            </exclusions>
           </dependency>
           <dependency>
             <groupId>org.apache.hbase</groupId>
             <artifactId>hbase-hadoop2-compat</artifactId>
             <version>${hbase.hadoop2.version}</version>
+            <type>test-jar</type>
+            <!-- hbase depends on hadoop-core which pollutes our classpath -->
+            <exclusions>
+              <exclusion>
+                <groupId>org.apache.hadoop</groupId>
+                <artifactId>hadoop-core</artifactId>
+              </exclusion>
+            </exclusions>
+          </dependency>
+          <dependency>
+            <groupId>org.apache.hbase</groupId>
+            <artifactId>hbase-server</artifactId>
+            <version>${hbase.hadoop2.version}</version>
+            <!-- hbase depends on hadoop-core which pollutes our classpath -->
+            <exclusions>
+              <exclusion>
+                <groupId>org.apache.hadoop</groupId>
+                <artifactId>hadoop-core</artifactId>
+              </exclusion>
+            </exclusions>
           </dependency>
           <dependency>
             <groupId>org.apache.hbase</groupId>
             <artifactId>hbase-server</artifactId>
             <version>${hbase.hadoop2.version}</version>
+            <type>test-jar</type>
+            <!-- hbase depends on hadoop-core which pollutes our classpath -->
+            <exclusions>
+              <exclusion>
+                <groupId>org.apache.hadoop</groupId>
+                <artifactId>hadoop-core</artifactId>
+              </exclusion>
+            </exclusions>
           </dependency>
           <dependency>
             <groupId>org.apache.hadoop</groupId>
diff --git a/ql/pom.xml b/ql/pom.xml
index 1e2eb20..d2326df 100644
--- a/ql/pom.xml
+++ b/ql/pom.xml
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
@@ -446,6 +446,12 @@
       <groupId>org.apache.spark</groupId>
       <artifactId>spark-core_${scala.binary.version}</artifactId>
       <version>${spark.version}</version>
+    </dependency>
+    <!-- Following have only one version, which is cdh.hadoop.version -->
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-common</artifactId>
+      <version>${hadoop-23.version}</version>
       <optional>true</optional>
     </dependency>
   </dependencies>
@@ -487,12 +493,6 @@
       <dependencies>
         <dependency>
           <groupId>org.apache.hadoop</groupId>
-          <artifactId>hadoop-common</artifactId>
-          <version>${hadoop-23.version}</version>
-          <optional>true</optional>
-        </dependency>
-        <dependency>
-          <groupId>org.apache.hadoop</groupId>
           <artifactId>hadoop-archives</artifactId>
           <version>${hadoop-23.version}</version>
           <optional>true</optional>
diff --git a/ql/src/java/org/apache/hadoop/hive/ql/exec/mr/JobDebugger.java b/ql/src/java/org/apache/hadoop/hive/ql/exec/mr/JobDebugger.java
index 6e4e3bf..8848485 100644
--- a/ql/src/java/org/apache/hadoop/hive/ql/exec/mr/JobDebugger.java
+++ b/ql/src/java/org/apache/hadoop/hive/ql/exec/mr/JobDebugger.java
@@ -116,8 +116,9 @@ public JobDebugger(JobConf conf, RunningJob rj, LogHelper console,
   public void run() {
     try {
       showJobFailDebugInfo();
-    } catch (IOException e) {
-      console.printError(e.getMessage());
+    } catch (Throwable t) {
+      t.printStackTrace(); // best we can do at this point
+      console.printError("Could not obtain debuging information: " + t.getClass().getSimpleName() + ": " + t.getMessage());
     }
   }
 
diff --git a/ql/src/java/org/apache/hadoop/hive/ql/io/IOContext.java b/ql/src/java/org/apache/hadoop/hive/ql/io/IOContext.java
index 04c9644..d51420d 100644
--- a/ql/src/java/org/apache/hadoop/hive/ql/io/IOContext.java
+++ b/ql/src/java/org/apache/hadoop/hive/ql/io/IOContext.java
@@ -53,7 +53,6 @@ private static IOContext get() {
    */
   private static final Map<String, IOContext> inputNameIOContextMap = new HashMap<String, IOContext>();
 
-
   public static IOContext get(Configuration conf) {
     if (HiveConf.getVar(conf, HiveConf.ConfVars.HIVE_EXECUTION_ENGINE).equals("spark")) {
       return get();
diff --git a/ql/src/java/org/apache/hadoop/hive/ql/parse/authorization/RestrictedHiveAuthorizationTaskFactoryImpl.java b/ql/src/java/org/apache/hadoop/hive/ql/parse/authorization/RestrictedHiveAuthorizationTaskFactoryImpl.java
new file mode 100644
index 0000000..4c7c7f8
--- /dev/null
+++ b/ql/src/java/org/apache/hadoop/hive/ql/parse/authorization/RestrictedHiveAuthorizationTaskFactoryImpl.java
@@ -0,0 +1,143 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hive.ql.parse.authorization;
+
+import java.io.Serializable;
+import java.util.HashSet;
+
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.ql.exec.Task;
+import org.apache.hadoop.hive.ql.hooks.ReadEntity;
+import org.apache.hadoop.hive.ql.hooks.WriteEntity;
+import org.apache.hadoop.hive.ql.metadata.Hive;
+import org.apache.hadoop.hive.ql.parse.ASTNode;
+import org.apache.hadoop.hive.ql.parse.SemanticException;
+
+/***
+ * Authorization factory that disables builtin authorization DDLs
+ */
+public class RestrictedHiveAuthorizationTaskFactoryImpl implements
+    HiveAuthorizationTaskFactory {
+
+  public RestrictedHiveAuthorizationTaskFactoryImpl(HiveConf conf, Hive db) {
+
+  }
+
+  @Override
+  public Task<? extends Serializable> createCreateRoleTask(ASTNode node,
+      HashSet<ReadEntity> inputs, HashSet<WriteEntity> outputs)
+      throws SemanticException {
+    raiseAuthError();
+    return null;
+  }
+
+  @Override
+  public Task<? extends Serializable> createDropRoleTask(ASTNode node,
+      HashSet<ReadEntity> inputs, HashSet<WriteEntity> outputs)
+      throws SemanticException {
+    raiseAuthError();
+    return null;
+  }
+
+  @Override
+  public Task<? extends Serializable> createShowRoleGrantTask(ASTNode node,
+      Path resultFile, HashSet<ReadEntity> inputs, HashSet<WriteEntity> outputs)
+      throws SemanticException {
+    raiseAuthError();
+    return null;
+  }
+
+  @Override
+  public Task<? extends Serializable> createGrantRoleTask(ASTNode node,
+      HashSet<ReadEntity> inputs, HashSet<WriteEntity> outputs)
+      throws SemanticException {
+    raiseAuthError();
+    return null;
+  }
+
+  @Override
+  public Task<? extends Serializable> createRevokeRoleTask(ASTNode node,
+      HashSet<ReadEntity> inputs, HashSet<WriteEntity> outputs)
+      throws SemanticException {
+    raiseAuthError();
+    return null;
+  }
+
+  @Override
+  public Task<? extends Serializable> createGrantTask(ASTNode node,
+      HashSet<ReadEntity> inputs, HashSet<WriteEntity> outputs)
+      throws SemanticException {
+    raiseAuthError();
+    return null;
+  }
+
+  @Override
+  public Task<? extends Serializable> createShowGrantTask(ASTNode node,
+      Path resultFile, HashSet<ReadEntity> inputs, HashSet<WriteEntity> outputs)
+      throws SemanticException {
+    raiseAuthError();
+    return null;
+  }
+
+  @Override
+  public Task<? extends Serializable> createRevokeTask(ASTNode node,
+      HashSet<ReadEntity> inputs, HashSet<WriteEntity> outputs)
+      throws SemanticException {
+    raiseAuthError();
+    return null;
+  }
+
+  @Override
+  public Task<? extends Serializable> createSetRoleTask(String roleName,
+      HashSet<ReadEntity> inputs, HashSet<WriteEntity> outputs)
+      throws SemanticException {
+    raiseAuthError();
+    return null;
+  }
+
+  @Override
+  public Task<? extends Serializable> createShowCurrentRoleTask(
+      HashSet<ReadEntity> inputs, HashSet<WriteEntity> outputs, Path resFile)
+      throws SemanticException {
+    raiseAuthError();
+    return null;
+  }
+
+  @Override
+  public Task<? extends Serializable> createShowRolesTask(ASTNode ast,
+      Path resFile, HashSet<ReadEntity> inputs, HashSet<WriteEntity> outputs)
+      throws SemanticException {
+    raiseAuthError();
+    return null;
+  }
+
+  @Override
+  public Task<? extends Serializable> createShowRolePrincipalsTask(ASTNode ast,
+      Path resFile, HashSet<ReadEntity> inputs, HashSet<WriteEntity> outputs)
+      throws SemanticException {
+    raiseAuthError();
+    return null;
+  }
+
+  private void raiseAuthError() throws SemanticException {
+    throw new SemanticException(
+        "The current builtin authorization in Hive is incomplete and disabled.");
+  }
+}
diff --git a/serde/pom.xml b/serde/pom.xml
index 795e3d2..df6f93f 100644
--- a/serde/pom.xml
+++ b/serde/pom.xml
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
@@ -109,6 +109,32 @@
       <groupId>com.sun.jersey</groupId>
       <artifactId>jersey-servlet</artifactId>
       <version>${jersey.version}</version>
+    </dependency>
+    <!-- Following have only one version, which is cdh.hadoop.version -->
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-common</artifactId>
+      <version>${hadoop-23.version}</version>
+      <optional>true</optional>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-common</artifactId>
+      <version>${hadoop-23.version}</version>
+      <classifier>tests</classifier>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-hdfs</artifactId>
+      <version>${hadoop-23.version}</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-hdfs</artifactId>
+      <version>${hadoop-23.version}</version>
+      <classifier>tests</classifier>
       <scope>test</scope>
     </dependency>
   </dependencies>
@@ -116,6 +142,9 @@
   <profiles>
     <profile>
       <id>hadoop-1</id>
+      <activation>
+        <activeByDefault>false</activeByDefault>
+      </activation>
       <dependencies>
         <dependency>
           <groupId>org.apache.hadoop</groupId>
@@ -136,36 +165,10 @@
       <dependencies>
         <dependency>
           <groupId>org.apache.hadoop</groupId>
-          <artifactId>hadoop-common</artifactId>
-          <version>${hadoop-23.version}</version>
-          <optional>true</optional>
-        </dependency>
-        <dependency>
-          <groupId>org.apache.hadoop</groupId>
-          <artifactId>hadoop-common</artifactId>
-          <version>${hadoop-23.version}</version>
-          <classifier>tests</classifier>
-          <scope>test</scope>
-        </dependency>
-        <dependency>
-          <groupId>org.apache.hadoop</groupId>
           <artifactId>hadoop-mapreduce-client-core</artifactId>
           <version>${hadoop-23.version}</version>
           <optional>true</optional>
         </dependency>
-        <dependency>
-          <groupId>org.apache.hadoop</groupId>
-          <artifactId>hadoop-hdfs</artifactId>
-          <version>${hadoop-23.version}</version>
-          <scope>test</scope>
-        </dependency>
-        <dependency>
-          <groupId>org.apache.hadoop</groupId>
-          <artifactId>hadoop-hdfs</artifactId>
-          <version>${hadoop-23.version}</version>
-          <classifier>tests</classifier>
-          <scope>test</scope>
-        </dependency>
       </dependencies>
     </profile>
   </profiles>
diff --git a/service/pom.xml b/service/pom.xml
index 44871bf..166d267 100644
--- a/service/pom.xml
+++ b/service/pom.xml
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
@@ -106,11 +106,21 @@
       <version>${junit.version}</version>
       <scope>test</scope>
     </dependency>
+    <!-- Following have only one version, which is cdh.hadoop.version -->
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-common</artifactId>
+      <version>${hadoop-23.version}</version>
+      <optional>true</optional>
+    </dependency>
   </dependencies>
 
   <profiles>
     <profile>
       <id>hadoop-1</id>
+      <activation>
+        <activeByDefault>false</activeByDefault>
+      </activation>
       <dependencies>
         <dependency>
           <groupId>org.apache.hadoop</groupId>
@@ -125,12 +135,6 @@
       <dependencies>
         <dependency>
           <groupId>org.apache.hadoop</groupId>
-          <artifactId>hadoop-common</artifactId>
-          <version>${hadoop-23.version}</version>
-          <optional>true</optional>
-        </dependency>
-        <dependency>
-          <groupId>org.apache.hadoop</groupId>
           <artifactId>hadoop-mapreduce-client-core</artifactId>
           <version>${hadoop-23.version}</version>
           <optional>true</optional>
diff --git a/shims/0.20S/pom.xml b/shims/0.20S/pom.xml
index 819fbbb..70493c8 100644
--- a/shims/0.20S/pom.xml
+++ b/shims/0.20S/pom.xml
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../../pom.xml</relativePath>
   </parent>
 
@@ -43,13 +43,20 @@
     <!-- inter-project -->
     <dependency>
       <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-common</artifactId>
+      <version>${hadoop-23.version}</version>
+      <optional>true</optional>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
       <artifactId>hadoop-core</artifactId>
       <version>${hadoop-20S.version}</version>
+      <classifier>tests</classifier>
       <optional>true</optional>
     </dependency>
     <dependency>
       <groupId>org.apache.hadoop</groupId>
-      <artifactId>hadoop-test</artifactId>
+      <artifactId>hadoop-core</artifactId>
       <version>${hadoop-20S.version}</version>
       <optional>true</optional>
     </dependency>
diff --git a/shims/0.23/pom.xml b/shims/0.23/pom.xml
index 7003248..b79de04 100644
--- a/shims/0.23/pom.xml
+++ b/shims/0.23/pom.xml
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../../pom.xml</relativePath>
   </parent>
 
diff --git a/shims/0.23/src/main/java/org/apache/hadoop/hive/shims/Hadoop23Shims.java b/shims/0.23/src/main/java/org/apache/hadoop/hive/shims/Hadoop23Shims.java
index 8c865ee..c4b7b75 100644
--- a/shims/0.23/src/main/java/org/apache/hadoop/hive/shims/Hadoop23Shims.java
+++ b/shims/0.23/src/main/java/org/apache/hadoop/hive/shims/Hadoop23Shims.java
@@ -74,7 +74,6 @@
 import org.apache.hadoop.mapred.RecordReader;
 import org.apache.hadoop.mapred.Reporter;
 import org.apache.hadoop.mapred.WebHCatJTShim23;
-import org.apache.hadoop.mapred.lib.TotalOrderPartitioner;
 import org.apache.hadoop.mapreduce.Job;
 import org.apache.hadoop.mapreduce.JobContext;
 import org.apache.hadoop.mapreduce.JobID;
@@ -158,8 +157,7 @@ public RecordReader getRecordReader(InputSplit split,
   public String getTaskAttemptLogUrl(JobConf conf,
     String taskTrackerHttpAddress, String taskAttemptId)
     throws MalformedURLException {
-    if (conf.get("mapreduce.framework.name") != null
-      && conf.get("mapreduce.framework.name").equals("yarn")) {
+    if (isMR2()) {
       // if the cluster is running in MR2 mode, return null
       LOG.warn("Can't fetch tasklog: TaskLogServlet is not supported in MR2 mode.");
       return null;
@@ -211,30 +209,49 @@ public TaskAttemptID newTaskAttemptID(JobID jobId, boolean isMap, int taskId, in
 
   @Override
   public boolean isLocalMode(Configuration conf) {
-    return "local".equals(conf.get("mapreduce.framework.name"));
+    if (isMR2()) {
+      return false;
+    }
+    return "local".equals(conf.get("mapreduce.framework.name")) ||
+      "local".equals(conf.get("mapred.job.tracker"));
   }
 
   @Override
   public String getJobLauncherRpcAddress(Configuration conf) {
-    return conf.get("yarn.resourcemanager.address");
+    if (isMR2()) {
+      return conf.get("yarn.resourcemanager.address");
+    } else {
+      return conf.get("mapred.job.tracker");
+    }
   }
 
   @Override
   public void setJobLauncherRpcAddress(Configuration conf, String val) {
     if (val.equals("local")) {
       // LocalClientProtocolProvider expects both parameters to be 'local'.
-      conf.set("mapreduce.framework.name", val);
-      conf.set("mapreduce.jobtracker.address", val);
+      if (isMR2()) {
+        conf.set("mapreduce.framework.name", val);
+        conf.set("mapreduce.jobtracker.address", val);
+      } else {
+        conf.set("mapred.job.tracker", val);
+      }
     }
     else {
-      conf.set("mapreduce.framework.name", "yarn");
-      conf.set("yarn.resourcemanager.address", val);
+      if (isMR2()) {
+        conf.set("yarn.resourcemanager.address", val);
+      } else {
+        conf.set("mapred.job.tracker", val);
+      }
     }
   }
 
   @Override
   public String getJobLauncherHttpAddress(Configuration conf) {
-    return conf.get("yarn.resourcemanager.webapp.address");
+    if (isMR2()) {
+      return conf.get("yarn.resourcemanager.webapp.address");
+    } else {
+      return conf.get("mapred.job.tracker.http.address");
+    }
   }
 
   protected boolean isExtendedAclEnabled(Configuration conf) {
@@ -259,7 +276,18 @@ public boolean moveToAppropriateTrash(FileSystem fs, Path path, Configuration co
 
   @Override
   public void setTotalOrderPartitionFile(JobConf jobConf, Path partitionFile){
-    TotalOrderPartitioner.setPartitionFile(jobConf, partitionFile);
+    try {
+      Class<?> clazz = Class.forName("org.apache.hadoop.mapred.lib.TotalOrderPartitioner");
+      try {
+        java.lang.reflect.Method method = clazz.getMethod("setPartitionFile", Configuration.class, Path.class);
+        method.invoke(null, jobConf, partitionFile);
+      } catch(NoSuchMethodException nsme) {
+        java.lang.reflect.Method method = clazz.getMethod("setPartitionFile", JobConf.class, Path.class);
+        method.invoke(null, jobConf, partitionFile);
+      }
+    } catch(Exception e) {
+      throw new IllegalStateException("Unable to find TotalOrderPartitioner.setPartitionFile", e);
+    }
   }
 
   @Override
@@ -341,6 +369,8 @@ public void shutdown() throws IOException {
     public void setupConfiguration(Configuration conf) {
       JobConf jConf = mr.createJobConf();
       for (Map.Entry<String, String> pair: jConf) {
+        // TODO figure out why this was wrapped in
+        // if(!"mapred.reduce.tasks".equalsIgnoreCase(pair.getKey()))
         conf.set(pair.getKey(), pair.getValue());
       }
     }
@@ -577,8 +607,15 @@ public JobContext createJobContext(Configuration conf,
     @Override
     public org.apache.hadoop.mapred.JobContext createJobContext(org.apache.hadoop.mapred.JobConf conf,
                                                                 org.apache.hadoop.mapreduce.JobID jobId, Progressable progressable) {
-      return new org.apache.hadoop.mapred.JobContextImpl(
-              new JobConf(conf), jobId, progressable);
+      try {
+        java.lang.reflect.Constructor construct = org.apache.hadoop.mapred.JobContextImpl.class.getDeclaredConstructor(
+          org.apache.hadoop.mapred.JobConf.class, org.apache.hadoop.mapreduce.JobID.class, Progressable.class);
+        construct.setAccessible(true);
+        return (org.apache.hadoop.mapred.JobContext) construct.newInstance(
+                new JobConf(conf), jobId, progressable);
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
     }
 
     @Override
@@ -600,13 +637,23 @@ public InetSocketAddress getResourceManagerAddress(Configuration conf) {
 
     @Override
     public String getPropertyName(PropertyName name) {
+      boolean mr2 = isMR2();
       switch (name) {
         case CACHE_ARCHIVES:
-          return MRJobConfig.CACHE_ARCHIVES;
+          if(mr2) {
+            return "mapreduce.job.cache.archives";
+          }
+          return "mapred.cache.archives";
         case CACHE_FILES:
-          return MRJobConfig.CACHE_FILES;
+          if(mr2) {
+            return "mapreduce.job.cache.files";
+          }
+          return "mapred.cache.files";
         case CACHE_SYMLINK:
-          return MRJobConfig.CACHE_SYMLINK;
+          if(mr2) {
+            return "mapreduce.job.cache.symlink.create";
+          }
+          return "mapred.create.symlink";
         case CLASSPATH_ARCHIVES:
           return MRJobConfig.CLASSPATH_ARCHIVES;
         case CLASSPATH_FILES:
@@ -791,6 +838,10 @@ public boolean apply(AclEntry input) {
   });
   }
 
+  private boolean isMR2() {
+    return org.apache.hadoop.mapred.MRVersion.isMR2();
+  }
+
   class ProxyFileSystem23 extends ProxyFileSystem {
     public ProxyFileSystem23(FileSystem fs) {
       super(fs);
diff --git a/shims/aggregator/pom.xml b/shims/aggregator/pom.xml
index 9d48a74..9d87a08 100644
--- a/shims/aggregator/pom.xml
+++ b/shims/aggregator/pom.xml
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../../pom.xml</relativePath>
   </parent>
 
@@ -39,12 +39,14 @@
       <version>${project.version}</version>
       <scope>compile</scope>
     </dependency>
+<!--
     <dependency>
       <groupId>org.apache.hive.shims</groupId>
       <artifactId>hive-shims-0.20S</artifactId>
       <version>${project.version}</version>
       <scope>runtime</scope>
     </dependency>
+-->
     <dependency>
       <groupId>org.apache.hive.shims</groupId>
       <artifactId>hive-shims-0.23</artifactId>
diff --git a/shims/common/pom.xml b/shims/common/pom.xml
index 5541543..08c49b6 100644
--- a/shims/common/pom.xml
+++ b/shims/common/pom.xml
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../../pom.xml</relativePath>
   </parent>
 
@@ -57,6 +57,12 @@
     </dependency>
     <dependency>
       <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-common</artifactId>
+      <version>${hadoop-23.version}</version>
+      <optional>true</optional>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
       <artifactId>hadoop-core</artifactId>
       <version>${hadoop-20S.version}</version>
       <optional>true</optional>
diff --git a/shims/pom.xml b/shims/pom.xml
index e72cffb..c672199 100644
--- a/shims/pom.xml
+++ b/shims/pom.xml
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
@@ -33,7 +33,9 @@
 
   <modules>
     <module>common</module>
+<!-- we do not use the 0.20S shim
     <module>0.20S</module>
+-->
     <module>0.23</module>
     <module>scheduler</module>
     <module>aggregator</module>
diff --git a/shims/scheduler/pom.xml b/shims/scheduler/pom.xml
index a1e0a21..516779e 100644
--- a/shims/scheduler/pom.xml
+++ b/shims/scheduler/pom.xml
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../../pom.xml</relativePath>
   </parent>
 
diff --git a/spark-client/pom.xml b/spark-client/pom.xml
index c00e912..3b89768 100644
--- a/spark-client/pom.xml
+++ b/spark-client/pom.xml
@@ -22,14 +22,13 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
   </parent>
 
   <groupId>org.apache.hive</groupId>
   <artifactId>spark-client</artifactId>
   <packaging>jar</packaging>
   <name>Spark Remote Client</name>
-  <version>1.1.0</version>
 
   <properties>
     <hive.path.to.root>..</hive.path.to.root>
diff --git a/testutils/pom.xml b/testutils/pom.xml
index 4af16c0..2279ac9 100644
--- a/testutils/pom.xml
+++ b/testutils/pom.xml
@@ -19,7 +19,7 @@
   <parent>
     <groupId>org.apache.hive</groupId>
     <artifactId>hive</artifactId>
-    <version>1.1.0</version>
+    <version>1.1.0-cdh5.4.0-SNAPSHOT</version>
     <relativePath>../pom.xml</relativePath>
   </parent>
 
-- 
1.7.9.5

